modules/core/src/main/java/org/apache/ignite/internal/visor/tx/VisorTxTask.java

```diff
@@ -173,21 +179,22 @@
                             it.remove();
                     }
                 }
+
+                if (infos.size() > limit)
+                    infos.subList(limit, infos.size()).clear();
             }
         }
 
         return mapRes;
     }
 
-    /**
-     *
-     */
+    /** */
     private static class VisorTxJob extends VisorJob<VisorTxTaskArg, VisorTxTaskResult> {
         /** */
         private static final long serialVersionUID = 0L;
 
         /** */
-        private static final int DEFAULT_LIMIT = 50;
+        private static final int DEFAULT_MAP_LIMIT = 50;
```

Comment:
The limit should be applied only on reducer side and removed from map side. Transactions should be streamed to reducer in batches using continous mapping to avoid memory issues.